File: user_interface/src/simulation/simulation.ts
export {}
// export function doEvent(value: EventJSON) {
//     request_data.invalid_action = value.type == "invalid_action" && !showQuestion.val;
//     requestData.decisionWarningText = "";
//     if (value.type == "request") {
//         showQuestion.val = true;
//         request_data.event = value;
//         request_data.questionPending = true;
//         switch_.val = !switch_.val;
//     } else if (value.type == "user_response") {
//         request_data.decision = request_data.event.choices[value.decision];
// 
//         request_data.questionPending = false;
//     } else if (value.type == "time_exceeded") {
//         requestData.decisionWarningText = "Took too long to take action. Used default choice";
//         request_data.decision = request_data.event.choices[value.default_choice];
//         request_data.questionPending = false;
//     } else if (value.type == "clear") {
//         showQuestion.val = false;
//     } else if (value.type == "invalid_action") {
//         
//         request_data.event.text = value.text;
//     }
// }
// 
// const clear_: EventJSON = {
//         type: "clear",
//         decision: 1,
//         choices: [],
//         default_choice: 0,
//         time_remaining: 0,
//         text: "",
//         request_text: ""
//     }
// var timeout: any = undefined;
// /**
//  * 
//  */
// export function startSimulation() {
//     const START_REQUEST: EventJSON = {
//         type: "request",
//         request_text: "Car in front is slow. Overtake when ready?",
//         choices: ["yes", "no"],
//         default_choice: 1, 
//         time_remaining: 4,
//         decision: 0,
//         text: ""
//     }
// 
//     const request_2: EventJSON = {
//         type: "request",
//         request_text: "I feel safe to take-over. Overtake now?",
//         choices: ["yes", "no"],
//         default_choice: 1, 
//         time_remaining: 3,
//         decision: 0,
//         text: ""
//     }
// 
//     const request_3: EventJSON = {
//         type: "request",
//         request_text: "Overtake now?",
//         choices: ["yes", "no"],
//         default_choice: 1, 
//         time_remaining: 5,
//         decision: 0,
//         text: ""
//     }
// 
//     const response_1: EventJSON = {
//         type: "user_response",
//         decision: 0,
//         choices: [],
//         default_choice: 0,
//         time_remaining: 0,
//         text: "",
//         request_text: ""
//     }
// 
//     
// 
//     const response_2: EventJSON = {
//         type: "user_response",
//         decision: 1,
//         choices: [],
//         default_choice: 0,
//         time_remaining: 0,
//         text: "",
//         request_text: ""
//     }
// 
//     const response_3: EventJSON = {
//         type: "user_response",
//         decision: 1,
//         choices: [],
//         default_choice: 0,
//         time_remaining: 0,
//         text: "",
//         request_text: ""
//     }
// 
//     timeout = setTimeout(() => {
//         doEvent(START_REQUEST);
// 
//         timeout = setTimeout(() => {
//             doEvent(response_1);
// 
//             timeout = setTimeout(() => {
//                 doEvent(clear_);
// 
//                 timeout = setTimeout(() => {
//                     doEvent(request_2);
// 
//                     timeout = setTimeout(() => {
//                         doEvent(response_2);
// 
//                         timeout = setTimeout(() => {
//                             doEvent(clear_);
//                         }, 1500);
//                     }, 2500)
//                 }, 2600); // i feel safe to take over
//             }, 1500)
//         }, 2500);
//     }, 3000);
// }

File: simulation.py

"""
this script simulates a real-life scenario to test the prototype
"""

import socket
import json
import logging
from configparser import ConfigParser
import time

CONFIG_FILE_PATH = 'config.ini'

config = ConfigParser()
config.read(CONFIG_FILE_PATH)

if not config.items:
    print('could not access config parameters')
    exit(1)

HOST = str(config['MAIN_SERVER']['HOST'])
REQUEST_PORT = int(config['MAIN_SERVER']['HOST_REQUEST_PORT'])
RESPONSE_PORT = int(config['MAIN_SERVER']['HOST_RESPONSE_PORT'])


def send_data(host, port, data):

    logging.info('Sending to: {}:{}'.format(host, port))
    logging.info(data)
    
    try: 
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # UDP

        # serialize json
        data = json.dumps(data)

        s.sendto(bytes(data, "utf-8"), (host, port))

    finally:
        s.close()
        logging.info('Succesfully sent request')

def main():

    logging.basicConfig(level=logging.INFO)

    request_1 = {
        "type": "start_simulation"
    }

    request_2 = {
        "type": "request",
        "request_text": "The car in front is driving slow. Take over when possible?",
        "choices": ["no", "yes"],
        "default_choice": 0, 
        "time_remaining": 5
    }

    request_3 = {
        "type": "request",
        "request_text": "I feel safe to take-over. Overtake now?",
        "choices": ["no", "yes"],
        "default_choice": 0, 
        "time_remaining": 5
    }

    request_4 = {
        "type": "reset_simulation"
    }

    # start simulation
    send_data(HOST, REQUEST_PORT, request_1)

    # ask for overtake
    time.sleep(3)
    send_data(HOST, REQUEST_PORT, request_2)

    # ask for overtake again
    time.sleep(7)
    send_data(HOST, REQUEST_PORT, request_3)

    # reset simulation
    time.sleep(20)
    send_data(HOST, REQUEST_PORT, request_4)


if __name__ == '__main__':
    main()